### PR TITLE
PR count notifications

### LINKED ISSUE #2296: Check open (undrafted) PR count per contributor and send a Slack notification prompting alternative tasks instead of new (reviewable) PRs that could be done
## Problem

Openverse contributors tend to highly prioritise code contributions over other types of contributions. Part of this comes from the enthusiasm we have for the improvements we can make to the project. Part of this also comes, I believe, from a lack of visibility, information, and expectations for other types of tasks that can be worked on that do not necessarily result in a new immediate code contribution. Unfortunately, this also leads to some of these tasks either being completely ignored by the entire team or being unevenly distributed across the core maintainers, neither of which are good for the overall health of the project or individual maintainers.

## Description

Add a new workflow triggered on PR open to count the number of reviewable[^reviewable] PRs currently open from the contributor opening a new PR. If it equals or exceeds 3 (see below "why three") and the contributor is present in the GitHub ID to Slack ID mapping, send them a pre-formatted message along the following lines:

```md
Howdy <contributor>!

You currently have <x> number of PRs open with requested reviews (totalling <x * 2> required reviews). To help ease the review burden, increase review velocity for older PRs, and improve the equitable distribution of project maintenance tasks across the team, please consider reviewing [this list of alternative project tasks](<link to docs page>) instead of working on new code contributions.
```

The number of open reviews should be dynamic in case someone uses their discretion to open PRs beyond the baseline number. We agree that this should be allowed as some situations will call for it (critical issues that you can solve quickly, etc). However, the reminder should still go out. Including the number of requested reviews helps emphasise how significant the review burden is each time a new PR is opened, especially compared to the capacity that reviewers will have to accomplish their own work.

[^reviewable]: Open and un-drafted. Reviewers are only pinged to review PRs that are "ready for review", i.e., not drafts. We should be able to create draft PRs and have them ready to mark "ready for review" once our PRs in need of review are reviewed.

### Why three?

We have a limited number of reviewers available for each PR. Each PR requires 2 approvals. For simplicity's sake, let's assume a consistent pool of 7 available reviewers and regular code contributors for the project. If each person has 3 PRs up for review, that would be 24 PRs. That would mean each reviewer, at any given time, would have just about 7 PRs assigned for them to review at a time:

2 required reviewers * 24 PRs = 48 total required reviews
48 reviews / 7 available reviewers = 6.86 reviews per person

That already seems like a lot, but the reality is that not everyone will be opening so many PRs to review at a time. However, because the pool of reviewers is not evenly spread across the different stacks and because some stacks tend to have a higher number of smaller PRs (which can often nonetheless still take significant time to review), I think 3 is appropriate. In any case, this issue should be implemented in such a way that this number can be easily adjusted up or down depending on the experience of the team after some time.

## Potential future improvements

Weigh the number of reviewable PRs for each contributor against the stack labels for those PRs. Because our reviewers are distributed unevenly across the app, it could make sense to prompt people to "either do one of these maintenance tasks or consider picking up an issue in a different part of the stack".

### PR DESCRIPTION
**Pull request by zackkrida**
## Fixes

Fixes #2296 by @sarayourfriend 

## Description

This action counts a PR author's open pull requests whenever they open a new PR. If the author has a slack uuid in our Slack-GitHub mapping and they have 3 or more open PRs, they will receive the following DM:

CleanShot 2023-06-08 at 13 10 51@2x

The script does not include draft PRs, or PRs with the following labels:

- '🤖 aspect: text'
- '🧱 stack: documentation'
- 'priority: critical'

Please let me know if there are other labels we should include!

## Testing Instructions

To test this action it was setup to run on push, so every time I commit a change to this PR. This will be removed before it is deployed. You can see the output of a recent runs here:

- Run when I had >=3 PRs: https://github.com/WordPress/openverse/actions/runs/5213895498
- Run when I had 2 open PRs: https://github.com/WordPress/openverse/actions/runs/5223279909

## Checklist

- [ ] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [ ] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by krysal** [comment]
It would be nice to exclude PRs labeled `aspect: text` or `stack: documentation` from the count as these are non-code work and very appreciated improvement to the docs instead!

File: automations/js/src/count_user_reviewable_prs.js
Comment (sarayourfriend) [inline automations/js/src/count_user_reviewable_prs.js:63]: If we move this check for `slackUsername` truthiness to the top, we can avoid querying.

**Comment by sarayourfriend** [comment]
I'm working on https://github.com/WordPress/openverse/issues/2304 today to help unblock this.

**Comment by sarayourfriend** [comment]
Actually, @zackkrida, I think we can link to https://docs.openverse.org/general/contributing.html#non-code-contributions in the meantime. I think https://github.com/WordPress/openverse/issues/2304 should be implemented as an expansion of that list rather than as an entirely new page. In any case, the documentation link in the notification can be updated after the fact if we do add a new page or section on an existing page for this.

**Comment by sarayourfriend** [comment]
I've opened https://github.com/WordPress/openverse/pull/2323 to add the new documentation page and expand on the existing ones. If this gets merged before that one then I'll update the reference in the notification. Otherwise, we can link to the maintainer tasks page in the notification in this PR.

**Comment by zackkrida** [comment]
Added the label filtering and made the slack username check earlier. I should be able to undraft this tomorrow with the real working Slack webhook!

**Comment by zackkrida** [comment]
I actually like that as-written, this will re-ping someone every single time they open a PR past the limit 👿

**Comment by sarayourfriend** [comment]
@zackkrida I think that's fine too, especially with the excellent idea to exclude certain labels that Krystle requested. In particular excluding critical issues makes a lot of sense, we should in no way make it annoying for folks to open those PRs as soon as they're able to.

File: .github/workflows/pr_limit_reminders.yml
Comment (sarayourfriend) [inline .github/workflows/pr_limit_reminders.yml:37]:
```suggestion
          msg="Howdy $slack_username! \nYou currently have $pr_count number of PRs open with requested reviews (totalling $((pr_count * 2)) required reviews). To help ease the review burden, increase review velocity for older PRs, and improve the equitable distribution of project maintenance tasks across the team, please consider reviewing [this list of ways to contribute](https://docs.openverse.org/meta/maintainer_tasks.html) instead of working on new code contributions."
```

The other PR is merged, so we can update the link here :slightly_smiling_face:

File: .github/workflows/pr_limit_reminders.yml
Comment (zackkrida) [inline .github/workflows/pr_limit_reminders.yml:5]:
```suggestion
    types: [opened, reopened]
```

File: .github/workflows/pr_limit_reminders.yml
Comment (zackkrida) [inline .github/workflows/pr_limit_reminders.yml:8]: Will be removed prior to merge.

**Review by sarayourfriend — APPROVED** [review]
LGTM :tada:

File: automations/js/src/count_user_reviewable_prs.js
Comment (krysal) [inline automations/js/src/count_user_reviewable_prs.js:49]:
```suggestion
    '🟥 priority: critical',
```

File: .github/workflows/pr_limit_reminders.yml
Comment (obulat) [inline .github/workflows/pr_limit_reminders.yml:5]:
```suggestion
    types: [opened, reopened, ready_for_review]
```
This is when a draft PR is marked "ready for review" - so the number of "open" PRs increases.

**Review by obulat — APPROVED** [review]
Looking forward to [not] getting messages from our bot :)

**Review by obulat — COMMENTED** [review]
It would've been nice to also exclude approved PRs from the PR count.

**Comment by zackkrida** [comment]
@obulat that's a good idea for which I'll make a follow-up issue.

**Pull request merged**
